### PR TITLE
Reworked how attributes work.

### PR DESCRIPTION
**Pull request by iwoplaza**

### CODE REVIEW

File: packages/typegpu/src/data/array.ts [hunk 1]
```diff
@@ -121,50 +118,53 @@ class TgpuLooseArrayImpl<TElement extends AnyTgpuData>
...
+    output.seekTo(beginning + this.size);
   }
 
   read(input: TB.ISerialInput): Parsed<Unwrap<TElement>>[] {
+    alignIO(input, this.byteAlignment);
     const elements: Parsed<Unwrap<TElement>>[] = [];
     for (let i = 0; i < this.elementCount; i++) {
-      if (this.elementType.isCustomAligned) {
-        alignIO(input, this.elementType.byteAlignment);
+      const customAlignment = getCustomAlignment(this.elementType);
```
Comment (mhawryluk) [inline packages/typegpu/src/data/array.ts]: custom alignment is the same for every element in the array, so there's no need to invoke this function for each of them. I think we can just use the value stored in `this.byteAlignment`

File: packages/typegpu/src/data/attributes.ts [hunk 2]
```diff
@@ -0,0 +1,335 @@
...
+}
+
+export interface Size<T extends number> {
+  type: 'size';
+  size: T;
+}
+
+export type AnyAttribute = Align<number> | Size<number>;
+
+export type AnyAttributeTuple = [...AnyAttribute[]];
```
Comment (mhawryluk) [inline packages/typegpu/src/data/attributes.ts]: doesn't that evaluate to just `AnyAttribute[]`? what's the difference?

**Review by mhawryluk — APPROVED** [review]
Love it! 🤩

File: packages/typegpu/src/data/attributes.ts [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+    if (this.size <= 0) {
+      throw new Error(
+        `Custom data size must be a positive number. Got: ${this.size}.`,
+      );
+    }
+  }
+
+  resolveReferences(ctx: IRefResolver): void {
+    throw new RecursiveDataTypeError();
+  }
```
Comment (reczkok) [inline packages/typegpu/src/data/attributes.ts]:
```suggestion
  resolveReferences(_: IRefResolver): void {
    throw new RecursiveDataTypeError();
  }
```

File: packages/typegpu/src/data/array.ts [hunk 4]
```diff
@@ -100,6 +96,7 @@ class TgpuArrayImpl<TElement extends AnyTgpuData>
       alignIO(input, this.byteAlignment);
       elements.push(this.elementType.read(input) as Parsed<Unwrap<TElement>>);
     }
+    alignIO(input, this.byteAlignment);
```
Comment (iwoplaza) [inline packages/typegpu/src/data/array.ts:99]: Aligning here as well, to make it match with how the array is measured.

File: packages/typegpu/src/data/attributes.ts (hunk 2 above)
Comment (mhawryluk) [reply packages/typegpu/src/data/attributes.ts]: I meant `type AnyAttributeTuple = AnyAttribute[];`
it looks to yield the same results, but I may be missing something

Comment (mhawryluk) [reply packages/typegpu/src/data/attributes.ts]: or AnyAttribute[] directly instead of [...AnyAttribute[]]

Comment (iwoplaza) [reply packages/typegpu/src/data/attributes.ts]: Good point, seems to work fine!

**Review by reczkok — APPROVED** [review]
Great idea, amazing implementation 🙇‍♂️🪿

**Pull request merged**